google/resource_compute_instance.go

```diff
@@ -542,41 +537,20 @@ func resourceComputeInstance() *schema.Resource {
 	}
 }
 
-func getInstance(config *Config, d *schema.ResourceData) (*computeBeta.Instance, error) {
-	computeApiVersion := getComputeApiVersion(d, InstanceBaseApiVersion, InstanceVersionedFeatures)
+func getInstance(config *Config, d *schema.ResourceData) (*compute.Instance, error) {
 	project, err := getProject(d, config)
 	if err != nil {
 		return nil, err
 	}
-
 	zone := d.Get("zone").(string)
-	instance := &computeBeta.Instance{}
-	switch computeApiVersion {
-	case v1:
-		instanceV1, err := config.clientCompute.Instances.Get(
-			project, zone, d.Id()).Do()
-		if err != nil {
-			return nil, handleNotFoundError(err, d, fmt.Sprintf("Instance %s", d.Get("name").(string)))
-		}
-
-		err = Convert(instanceV1, instance)
-		if err != nil {
-			return nil, err
-		}
-	case v0beta:
-		var err error
-		instance, err = config.clientComputeBeta.Instances.Get(
-			project, zone, d.Id()).Do()
-		if err != nil {
-			return nil, handleNotFoundError(err, d, fmt.Sprintf("Instance %s", d.Get("name").(string)))
-		}
+	instance, err := config.clientCompute.Instances.Get(project, zone, d.Id()).Do()
```

Comment:
I want to think through what that would look like for a resource that supported the beta API but that didn't currently use any beta features.

If I understand correctly, using this resource as an example:
* There would be an empty `InstanceVersionedFeatures`, i.e. `var InstanceVersionedFeatures = []Feature{}`
* We'd keep all the `switch computeApiVersion` blocks and keep up-converting `compute` structs to their `computeBeta` counterparts.
* All flatten/expand helper functions (which are potentially shared by multiple resources) would take `computeBeta` arguments, regardless of whether they contained any beta functionality.

I think that could work. It would be a little confusing that everything took `computeBeta` arguments regardless of whether it actually contained any beta functionality, and would be especially unfortunate if we needed to keep maintaining two implementations of each helper function.

I'm on the fence and happy to defer to the provider's maintainers. I can put the scaffolding for beta resources back in. I think the code is easier to reason about without the beta resources, but empathise that someone will likely have to add all the boilerplate back next time we want to support a beta feature on this resource.

Reply:
I was thinking about this more this morning. I'm increasingly feeling that it's more readable to leave out beta support for resources that don't make use of beta features, assuming that re-adding beta support is as simple as:
1. Re-inserting well-established boilerplate (`InstanceVersionedFeatures`, `switch computeApiVersion`)
2. `s/compute/computeBeta/g`.

Reply:
I think it's more readable, but also less maintainable, if that makes sense. Right now, the barrier to entry of using a versioned feature is adding it to the `InstanceVersionedFeatures` list. If we remove all the current beta scaffolding, then the barrier to entry becomes adding it all back. If we get to a point where we don't have any beta features in any resources, then we would need to go back into the files' histories to find examples of how to convert back and forth. Or, a new contributor might not even realize that we have examples already, and then they'd have to try to figure it all out again. For a user adding something unrelated to beta features, the extra scaffolding is easy to work around, but if the code isn't there it could potentially become very difficult for someone to figure out how to add it back. So I stand by my original opinion that we should leave it in.

Reply:
@danawillow Fair enough. How do you think I should proceed here?

I believe that if I re-add the beta scaffolding to `resource_compute_instance` I'll need to switch all of its helper functions to take/produce `computeBeta` structs.

A big part of this PR's intent was to reuse code between `resource_compute_instance` and `resource_compute_instance_template` in order to make the two resources more consistent from a Terraform user's perspective. If I reintroduce the beta scaffolding for `resource_compute_instance` I'll need to also add said scaffolding to `resource_compute_instance_template` in order for the two instances to use the same code.

I'm happy to add beta scaffolding to both resources, but I'm also open to abandoning this PR and sending a new one that adds what I really want (Alias IP support for instance templates) if you think this is all getting too unwieldy.

Reply:
Sure, but if you switch it to non-beta and then a beta feature comes along, you still would need to figure out what to do about beta-ifying instance template.

I'm ok with adding the Alias IP support in a new PR so we can keep having the refactoring conversations in this one while that gets in (it'll probably be a quick review since it'll have mostly already been reviewed)